tools/koch/builder.nim

```diff
@@ -0,0 +1,438 @@
+  ## Add `inputs` and return the first ID
+  if b.inputs.high + inputs.len > high(Input).int:
+    raise newException(ValueError):
+      "Adding inputs will exceed the maximum capacity of " & $high(Input)
+
+  result = Input b.inputs.len
+  b.inputs.add inputs
+
+func addExtraOperands(b: var Builder, operands: varargs[Operand]): Operand =
+  ## Add `operands` to the list of extra operands and return the index of the
+  ## first operand added as an operand
+  if b.extraOperands.high + operands.len > high(Operand).int:
+    raise newException(ValueError):
+      "Adding operands will exceed the maximum capacity of " & $high(Operand)
+
+  result = Operand b.extraOperands.len
+  b.extraOperands.add operands
+
+func topologicalOrder(b: Builder): seq[Target] =
+  ## Produce the topological ordering of all targets in the builder
+  func pop[T](s: var set[T]): T =
+    ## Pop an item from the set
+    for item in s.items:
+      result = item
+      break
+
+    s.excl(result)
+
+  # Don't do anything if there are no targets
+  if b.len == 0: return
+
+  var
+    starters = {Target(0) .. Target(b.len - 1)}
+      ## List of targets without dependencies
+    dependencies = b.dependencies
+      ## The target dependency table
+
+  # Exclude all targets with dependencies from starters list
+  for targetWithDep in dependencies.keys:
+    starters.excl targetWithDep
+
+  # Produce a topological ordering using Kahn's algorithm
+  while starters != {}:
+    let target = starters.pop()
+    result.add target
+
+    for targetWithDep, deps in dependencies.mpairs:
+      if target in deps:
+        deps.excl target
+
+        if deps == {}:
+          starters.incl targetWithDep
+
+  # Why is this an assert?
+  #
+  # The graph is constructed by declaring a node and its dependency at the same
+  # time, so it is impossible for a later node to be referenced by a previous
+  # one.
+  #
+  # If this happens, we messed up the code.
+  for deps in dependencies.values:
+    assert deps == {}:
+      "There is a cycle in the dependency graph"
+
+func sortTargets(b: var Builder): seq[Target] =
+  ## Perform a topological sort on builders' targets.
+  let sortedOrder = b.topologicalOrder
+
+  var
+    newNames: typeof(b.names)
+    newRecipes: typeof(b.recipes)
+    newDependencies: typeof(b.dependencies)
+
+  for idx, target in sortedOrder.pairs:
+    # Insert to new target seq according to the sorted order
+    newRecipes.add b.recipes[target]
+    newNames.add b.names[target]
+
+    # If the target has any dependencies
+    let deps = b.dependencies.getOrDefault(target)
+    if deps != {}:
+      for newDep in Target(0) ..< target:
+        let dep = sortedOrder[newDep]
+        # If dep was a dependency of the target
+        if dep in deps:
+          # Re-link it to the new ID
+          newDependencies[target].incl newDep
+
+  b.names = newNames
+  b.recipes = newRecipes
+  b.dependencies = newDependencies
+
+func name*(b: Builder, target: Target): string =
```

Comment:
This is where it might be nice to have `Target` be `TargetId` and `Target` be:
```nim
type
  Target = object
    builder: Builder
    id: TargetId
```

at least from the outside people are dealing with a friendlier concept of a holistic target, while internally the memory layout and data are nice and orderly.

Reply:
It will be at one point, but having builder as a `ref` is not something I'd like now due to how easy it is to hide mutations.

Reply:
Well it could only happen inside module defined procs and if you wanted extra safety the builder could be a distinct to reduce the surface further.